// @tsed/platform-params@7.8.1/lib/cjs/decorators/bodyParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RawBodyParams = exports.BodyParams = void 0;
const ParamTypes_1 = require("../domain/ParamTypes");
const useParam_1 = require("./useParam");
const mapParamsOptions_1 = require("../utils/mapParamsOptions");
function BodyParams(...args) {
    const { expression, useType, useMapper = true, useValidation = true } = (0, mapParamsOptions_1.mapParamsOptions)(args);
    return (0, useParam_1.UseParam)({
        paramType: ParamTypes_1.ParamTypes.BODY,
        expression,
        useType,
        useMapper,
        useValidation
    });
}
exports.BodyParams = BodyParams;
/**
 * RawBodyParams return the value from [request.body](http://expressjs.com/en/4x/api.html#req.body) as a Buffer.
 *
 * #### Example
 *
 * ```typescript
 * @Controller('/')
 * class MyCtrl {
 *    @Post('/')
 *    create(@RawBodyParams() body: Buffer) {
 *       console.log('Entire body', body.toString("utf8"));
 *    }
 * }
 * ```
 *
 * @decorator
 * @operation
 * @input
 * @alias BodyParams Example: @BodyParams() payload: Buffer
 */
function RawBodyParams() {
    return (0, useParam_1.UseParam)({
        paramType: ParamTypes_1.ParamTypes.RAW_BODY,
        useMapper: false,
        useValidation: false
    });
}
exports.RawBodyParams = RawBodyParams;
