// @tsed/platform-params@7.8.1/lib/cjs/decorators/queryParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RawQueryParams = exports.QueryParams = void 0;
const ParamTypes_1 = require("../domain/ParamTypes");
const mapParamsOptions_1 = require("../utils/mapParamsOptions");
const useParam_1 = require("./useParam");
function QueryParams(...args) {
    const { expression, useType, useMapper = true, useValidation = true } = (0, mapParamsOptions_1.mapParamsOptions)(args);
    return (0, useParam_1.UseParam)({
        paramType: ParamTypes_1.ParamTypes.QUERY,
        expression,
        useType,
        useMapper,
        useValidation
    });
}
exports.QueryParams = QueryParams;
/**
 * RawQueryParams return the value from [request.query](http://expressjs.com/en/4x/api.html#req.query) object.
 *
 * Any validation and transformation are performed on the value. Use [pipes](/docs/pipes.html) to validate and/or transform the value.
 *
 * #### Example
 *
 * ```typescript
 * @Controller('/')
 * class MyCtrl {
 *    @Get('/')
 *    get(@RawPathParams() params: any) {
 *       console.log('Entire params', params);
 *    }
 *
 *    @Get('/')
 *    get(@RawPathParams('id') id: string) {
 *       console.log('ID', id);
 *    }
 * }
 * ```
 * > For more information on deserialization see [converters](/docs/json-mapper.md) page.
 *
 * @param expression The path of the property to get.
 * @decorator
 * @operation
 * @input
 */
function RawQueryParams(expression) {
    return QueryParams({
        expression,
        useConverters: false,
        useValidation: false
    });
}
exports.RawQueryParams = RawQueryParams;
