// @tsed/platform-params@7.8.1/lib/esm/decorators/bodyParams.js
import { ParamTypes } from "../domain/ParamTypes.js";
import { UseParam } from "./useParam.js";
import { mapParamsOptions } from "../utils/mapParamsOptions.js";
export function BodyParams(...args) {
    const { expression, useType, useMapper = true, useValidation = true } = mapParamsOptions(args);
    return UseParam({
        paramType: ParamTypes.BODY,
        expression,
        useType,
        useMapper,
        useValidation
    });
}
/**
 * RawBodyParams return the value from [request.body](http://expressjs.com/en/4x/api.html#req.body) as a Buffer.
 *
 * #### Example
 *
 * ```typescript
 * @Controller('/')
 * class MyCtrl {
 *    @Post('/')
 *    create(@RawBodyParams() body: Buffer) {
 *       console.log('Entire body', body.toString("utf8"));
 *    }
 * }
 * ```
 *
 * @decorator
 * @operation
 * @input
 * @alias BodyParams Example: @BodyParams() payload: Buffer
 */
export function RawBodyParams() {
    return UseParam({
        paramType: ParamTypes.RAW_BODY,
        useMapper: false,
        useValidation: false
    });
}
