// @tsed/cli-plugin-prisma@4.2.4/lib/cjs/hooks/PrismaInitHook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismaInitHook = void 0;
const tslib_1 = require("tslib");
const cli_core_1 = require("@tsed/cli-core");
const di_1 = require("@tsed/di");
const CliPrisma_1 = require("../services/CliPrisma");
let PrismaInitHook = class PrismaInitHook {
    cliPrisma;
    cliService;
    packageJson;
    async onExec(ctx) {
        this.addScripts();
        this.addDependencies(ctx);
        this.addDevDependencies(ctx);
        return [
            {
                title: "Generate Prisma schema",
                task: () => this.cliPrisma.init()
            },
            {
                title: "Add Ts.ED configuration to Prisma schema",
                task: () => this.cliPrisma.patchPrismaSchema()
            }
        ];
    }
    addScripts() {
        this.packageJson.addScripts({
            "prisma:migrate": "npx prisma migrate dev --name init",
            "prisma:generate": "npx prisma generate"
        });
    }
    addDependencies(ctx) {
        this.packageJson.addDependencies({
            "@tsed/prisma": "latest",
            "@prisma/client": "latest"
        }, ctx);
    }
    addDevDependencies(ctx) {
        this.packageJson.addDevDependencies({
            prisma: "latest"
        }, ctx);
    }
    $onFinish() {
        return new Promise((resolve) => {
            this.packageJson.runScript("prisma:generate").subscribe({
                complete() {
                    resolve([]);
                },
                error: () => {
                    resolve([]);
                }
            });
        });
    }
};
tslib_1.__decorate([
    (0, cli_core_1.Inject)(),
    tslib_1.__metadata("design:type", CliPrisma_1.CliPrisma)
], PrismaInitHook.prototype, "cliPrisma", void 0);
tslib_1.__decorate([
    (0, cli_core_1.Inject)(),
    tslib_1.__metadata("design:type", cli_core_1.CliService)
], PrismaInitHook.prototype, "cliService", void 0);
tslib_1.__decorate([
    (0, cli_core_1.Inject)(),
    tslib_1.__metadata("design:type", cli_core_1.ProjectPackageJson)
], PrismaInitHook.prototype, "packageJson", void 0);
tslib_1.__decorate([
    (0, cli_core_1.OnExec)("init"),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", Promise)
], PrismaInitHook.prototype, "onExec", null);
PrismaInitHook = tslib_1.__decorate([
    (0, di_1.Injectable)()
], PrismaInitHook);
exports.PrismaInitHook = PrismaInitHook;
