// @tsed/cli-core@4.2.4/lib/cjs/utils/getPackageJson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPackageJson = void 0;
const tslib_1 = require("tslib");
const read_pkg_up_1 = tslib_1.__importDefault(require("read-pkg-up"));
const path_1 = require("path");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
function useReadPkgUp(configuration) {
    return !(process.argv.includes("init") && !fs_extra_1.default.existsSync((0, path_1.join)(String(configuration.project?.rootDir), "package.json")));
}
function getEmptyPackageJson(configuration) {
    return {
        name: configuration.name,
        version: "1.0.0",
        description: "",
        scripts: {},
        dependencies: {},
        devDependencies: {}
    };
}
function getPackageJson(configuration) {
    if (useReadPkgUp(configuration)) {
        const result = read_pkg_up_1.default.sync({
            cwd: configuration.project?.rootDir
        });
        if (result && result.path) {
            const pkgPath = (0, path_1.dirname)(result.path);
            configuration.set("project.root", pkgPath);
            const pkg = fs_extra_1.default.readJsonSync(result.path, { encoding: "utf8" });
            return { ...getEmptyPackageJson(configuration), ...pkg };
        }
    }
    return getEmptyPackageJson(configuration);
}
exports.getPackageJson = getPackageJson;
