// @tsed/perf@7.8.1/lib/cjs/domain/Perf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Perf = void 0;
const core_1 = require("@tsed/core");
const utils_1 = require("../utils/utils");
// istanbul ignore next
class Perf {
    #latest;
    #start;
    constructor() {
        this.start = this.start.bind(this);
        this.end = this.end.bind(this);
        this.run = this.run.bind(this);
        this.fromStart = this.fromStart.bind(this);
        this.fromLatest = this.fromLatest.bind(this);
    }
    start() {
        this.#start = this.#latest = (0, utils_1.now)();
        return this;
    }
    async runFor(it, fn) {
        const { time } = await this.run(async () => {
            for (let i = 0; i < it; i++) {
                await fn();
            }
        });
        return time;
    }
    run(fn, onTime) {
        const date = (0, utils_1.now)();
        const result = fn();
        const getDiff = (result) => {
            const diff = (0, utils_1.fromNow)(date);
            this.#latest = (0, utils_1.now)();
            onTime && onTime(diff);
            return onTime ? result : { result, time: diff };
        };
        if ((0, core_1.isPromise)(result)) {
            return result.then(getDiff);
        }
        return getDiff(result);
    }
    fromStart() {
        return (0, utils_1.fromNow)(this.#start);
    }
    fromLatest() {
        const diff = (0, utils_1.fromNow)(this.#latest);
        this.#latest = (0, utils_1.now)();
        return diff;
    }
    end() {
        const diff = (0, utils_1.fromNow)(this.#start);
        this.#start = this.#latest = (0, utils_1.now)();
        return diff;
    }
}
exports.Perf = Perf;
