// @tsed/cli-core@4.2.4/lib/cjs/services/CliRunScript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliRunScript = void 0;
const tslib_1 = require("tslib");
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const run_script_1 = tslib_1.__importDefault(require("@npmcli/run-script"));
const di_1 = require("@tsed/di");
const ProjectPackageJson_1 = require("./ProjectPackageJson");
let CliRunScript = class CliRunScript {
    projectPackageJson;
    run(cmd, args, options = {}) {
        return (0, run_script_1.default)({
            event: "run",
            ...options,
            cmd: `${cmd} ${args.join(" ")}`,
            path: options.cwd || this.projectPackageJson.dir,
            env: options.env || {},
            stdio: options.stdio || "inherit",
            banner: false
        });
    }
};
tslib_1.__decorate([
    (0, di_1.Inject)(),
    tslib_1.__metadata("design:type", ProjectPackageJson_1.ProjectPackageJson)
], CliRunScript.prototype, "projectPackageJson", void 0);
CliRunScript = tslib_1.__decorate([
    (0, di_1.Injectable)()
], CliRunScript);
exports.CliRunScript = CliRunScript;
