// @tsed/cli-plugin-prisma@4.2.4/lib/cjs/commands/PrismaCmd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismaCmd = void 0;
const tslib_1 = require("tslib");
const cli_core_1 = require("@tsed/cli-core");
const CliPrisma_1 = require("../services/CliPrisma");
let PrismaCmd = class PrismaCmd {
    cli;
    async $exec(ctx) {
        return [
            {
                title: `Run Prisma CLI ${ctx.command}`,
                task: () => this.cli.run(ctx.command, ctx.rawArgs)
            }
        ];
    }
};
tslib_1.__decorate([
    (0, cli_core_1.Inject)(),
    tslib_1.__metadata("design:type", CliPrisma_1.CliPrisma)
], PrismaCmd.prototype, "cli", void 0);
PrismaCmd = tslib_1.__decorate([
    (0, cli_core_1.Command)({
        name: "prisma",
        description: "Run a prisma command",
        args: {
            command: {
                description: "The prisma command",
                type: String,
                required: true
            }
        },
        options: {},
        allowUnknownOption: true
    })
], PrismaCmd);
exports.PrismaCmd = PrismaCmd;
