// @tsed/engines@7.8.1/lib/cjs/utils/cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readPartials = exports.read = exports.cache = exports.clearCache = exports.getFromCache = exports.setToCache = exports.setEngine = exports.importEngine = exports.getCachedEngine = exports.requires = void 0;
const tslib_1 = require("tslib");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const path_1 = require("path");
const readCache = new Map();
const cacheStore = new Map();
exports.requires = new Map();
function getCachedEngine(name, toName = name) {
    const mdl = exports.requires.get(toName) || {};
    return mdl.default || mdl;
}
exports.getCachedEngine = getCachedEngine;
async function importEngine(name, toName = name) {
    if (!exports.requires.has(toName)) {
        await setEngine(name, toName);
    }
    return getCachedEngine(toName);
}
exports.importEngine = importEngine;
async function setEngine(name, toName = name) {
    const mdl = await Promise.resolve().then(() => tslib_1.__importStar(require(name)));
    exports.requires.set(toName, mdl);
}
exports.setEngine = setEngine;
function setToCache(key, value) {
    readCache.set(key, value);
}
exports.setToCache = setToCache;
function getFromCache(key) {
    return readCache.get(key);
}
exports.getFromCache = getFromCache;
/**
 * Clear the cache.
 *
 * @api public
 */
const clearCache = () => {
    readCache.clear();
    cacheStore.clear();
};
exports.clearCache = clearCache;
/**
 * Conditionally cache `compiled` template based
 * on the `options` filename and `.cache` boolean.
 *
 * @param {Object} options
 * @param {Function} compiled
 * @return {Function}
 */
function cache(options, compiled) {
    // cachable
    if (compiled && options.filename && options.cache) {
        readCache.delete(options.filename);
        cacheStore.set(options.filename, compiled);
        return compiled;
    }
    // check cache
    if (options.filename && options.cache) {
        return cacheStore.get(options.filename);
    }
    return compiled;
}
exports.cache = cache;
/**
 * Read `path` with `options` with
 * callback `(err, str)`. When `options.cache`
 * is true the template string will be cached.
 *
 * @param path
 * @param {String} options
 */
async function read(path, options) {
    let str = readCache.get(path);
    // cached (only if cached is a string and not a compiled template function)
    if (options.cache && str && typeof str === "string") {
        return str;
    }
    // read
    str = await fs_extra_1.default.readFile(path, "utf8");
    // remove extraneous utf8 BOM marker
    str = str.replace(/^\uFEFF/, "");
    if (options.cache) {
        readCache.set(path, str);
    }
    return str;
}
exports.read = read;
/**
 * Read `path` with `options` with
 * callback `(err, str)`. When `options.cache`
 * is true the partial string will be cached.
 *
 * @param path
 * @param {String} options
 * @api private
 */
async function readPartials(path, options) {
    if (!options.partials) {
        return;
    }
    const keys = Object.keys(options.partials);
    const partials = {};
    return new Promise((resolve, reject) => {
        async function next(index) {
            if (index === keys.length) {
                return resolve(partials);
            }
            const key = keys[index];
            const partialPath = options.partials[key];
            if (partialPath === undefined || partialPath === null || partialPath === false) {
                return next(++index);
            }
            let file;
            if ((0, path_1.isAbsolute)(partialPath)) {
                if ((0, path_1.extname)(partialPath) !== "") {
                    file = partialPath;
                }
                else {
                    file = (0, path_1.join)(partialPath + (0, path_1.extname)(path));
                }
            }
            else {
                file = (0, path_1.join)((0, path_1.dirname)(path), partialPath + (0, path_1.extname)(path));
            }
            try {
                partials[key] = await read(file, options);
                next(++index);
            }
            catch (err) {
                reject(err);
            }
        }
        next(0);
    });
}
exports.readPartials = readPartials;
