// @tsed/prisma@7.8.1/lib/esm/generator/domain/ScalarTsTypes.js
export var PrismaScalars;
(function (PrismaScalars) {
    PrismaScalars["String"] = "String";
    PrismaScalars["Boolean"] = "Boolean";
    PrismaScalars["Int"] = "Int";
    PrismaScalars["Float"] = "Float";
    PrismaScalars["BigInt"] = "BigInt";
    PrismaScalars["Decimal"] = "Decimal";
    PrismaScalars["DateTime"] = "DateTime";
    PrismaScalars["Json"] = "Json";
    PrismaScalars["Bytes"] = "Bytes";
})(PrismaScalars || (PrismaScalars = {}));
export const ScalarTsTypes = {
    [PrismaScalars.String]: "string",
    [PrismaScalars.Boolean]: "boolean",
    [PrismaScalars.Int]: "number",
    [PrismaScalars.Float]: "number",
    [PrismaScalars.Decimal]: "number",
    [PrismaScalars.BigInt]: "bigint",
    [PrismaScalars.DateTime]: "Date",
    [PrismaScalars.Json]: "any",
    [PrismaScalars.Bytes]: "Buffer"
};
export const ScalarJsClasses = {
    [PrismaScalars.String]: "String",
    [PrismaScalars.Boolean]: "Boolean",
    [PrismaScalars.Int]: "Number",
    [PrismaScalars.Float]: "Number",
    [PrismaScalars.Decimal]: "Number",
    [PrismaScalars.BigInt]: "BigInt",
    [PrismaScalars.DateTime]: "Date",
    [PrismaScalars.Json]: "Object",
    [PrismaScalars.Bytes]: "Buffer"
};
export const ScalarDecorators = {
    [PrismaScalars.String]: [],
    [PrismaScalars.Boolean]: [],
    [PrismaScalars.Int]: [
        {
            name: "Integer",
            arguments: []
        }
    ],
    [PrismaScalars.Float]: [],
    [PrismaScalars.BigInt]: [],
    [PrismaScalars.Decimal]: [],
    [PrismaScalars.DateTime]: [{ name: "Format", arguments: ['"date-time"'] }],
    [PrismaScalars.Json]: [],
    [PrismaScalars.Bytes]: []
};
