// @tsed/common@7.8.1/lib/esm/services/PlatformMiddlewaresChain.js
import { __decorate, __metadata } from "tslib";
import { isClass } from "@tsed/core";
import { Constant, Inject, Injectable } from "@tsed/di";
import { ParamTypes } from "@tsed/platform-params";
import { JsonEntityStore } from "@tsed/schema";
import { PlatformAcceptMimesMiddleware } from "../middlewares/PlatformAcceptMimesMiddleware.js";
import { PlatformMulterMiddleware } from "../middlewares/PlatformMulterMiddleware.js";
import { PlatformAdapter } from "../services/PlatformAdapter.js";
let PlatformMiddlewaresChain = class PlatformMiddlewaresChain {
    acceptMimes;
    adapter;
    get(allMiddlewares, operationRoute) {
        const { ACCEPT_MIMES, FILE } = this.getParamTypes(allMiddlewares, operationRoute);
        return [ACCEPT_MIMES && PlatformAcceptMimesMiddleware, FILE && PlatformMulterMiddleware, ...allMiddlewares].filter(Boolean);
    }
    hasAcceptMimes(operationRoute) {
        return operationRoute.endpoint.acceptMimes.length || this.acceptMimes.length;
    }
    getParamTypes(middlewares, operationRoute) {
        return middlewares.filter(isClass).reduce((paramTypes, token) => {
            if (token !== operationRoute.endpoint) {
                const entity = JsonEntityStore.fromMethod(token, "use");
                if (entity.decoratorType === "method") {
                    const { FILE, RAW_BODY, BODY } = entity.getParamTypes();
                    paramTypes.FILE = paramTypes.FILE || FILE;
                    paramTypes.RAW_BODY = paramTypes.RAW_BODY || RAW_BODY;
                    paramTypes.BODY = paramTypes.BODY || BODY;
                }
            }
            return paramTypes;
        }, {
            ACCEPT_MIMES: this.hasAcceptMimes(operationRoute),
            FILE: operationRoute.has(ParamTypes.FILES),
            RAW_BODY: operationRoute.has(ParamTypes.RAW_BODY),
            BODY: operationRoute.has(ParamTypes.BODY) || operationRoute.method === "ALL"
        });
    }
};
__decorate([
    Constant("acceptMimes", []),
    __metadata("design:type", Array)
], PlatformMiddlewaresChain.prototype, "acceptMimes", void 0);
__decorate([
    Inject(),
    __metadata("design:type", PlatformAdapter)
], PlatformMiddlewaresChain.prototype, "adapter", void 0);
PlatformMiddlewaresChain = __decorate([
    Injectable()
], PlatformMiddlewaresChain);
export { PlatformMiddlewaresChain };
