// @tsed/platform-router@7.8.1/lib/cjs/domain/PlatformLayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformLayer = void 0;
const core_1 = require("@tsed/core");
class PlatformLayer {
    provider;
    path = "";
    method;
    handlers = [];
    router;
    opts = {};
    #args;
    constructor(props = {}) {
        Object.assign(this, props);
    }
    set(args) {
        this.#args = args;
    }
    getArgs() {
        return [this.path, ...this.#args].filter(Boolean);
    }
    isProvider() {
        return !!this.provider;
    }
    inspect() {
        return {
            path: this.path,
            method: this.method,
            handlers: this.handlers.map((item) => String(item)),
            opts: Object.entries(this.opts).reduce((obj, [key, value]) => {
                return {
                    ...obj,
                    [key]: (0, core_1.isPrimitive)(value) ? value : (0, core_1.nameOf)(value)
                };
            }, {})
        };
    }
}
exports.PlatformLayer = PlatformLayer;
