// @tsed/prisma@7.8.1/lib/cjs/generator/domain/DmmfModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DmmfModel = void 0;
const core_1 = require("@tsed/core");
const ts_morph_1 = require("ts-morph");
const DmmfField_1 = require("./DmmfField");
const change_case_1 = require("change-case");
class DmmfModel {
    isInputType;
    model;
    modelType;
    #imports = new Map();
    constructor({ isInputType, model, modelType }) {
        this.model = model;
        this.modelType = modelType;
        this.isInputType = isInputType;
    }
    get name() {
        return this.model.name;
    }
    get fields() {
        const dataField = (0, core_1.toMap)(this.modelType.fields, "name");
        return this.model.fields.map((field) => {
            return new DmmfField_1.DmmfField({
                model: this,
                field,
                schemaArg: dataField.get(field.name)
            });
        });
    }
    get importDeclarations() {
        return [...this.#imports.values()];
    }
    static getModels(dmmf, modelsMap, typesMap) {
        const inputObjectsTypes = dmmf.schema.inputObjectTypes.model || [];
        const outputObjectTypes = dmmf.schema.outputObjectTypes.model || [];
        const factory = (isInputType) => (modelType) => {
            let model = modelsMap.get(modelType.name) || typesMap?.get(modelType.name);
            if (model) {
                return new DmmfModel({
                    modelType,
                    model,
                    isInputType
                });
            }
            return undefined;
        };
        const inputs = inputObjectsTypes.map(factory(true)).filter(Boolean);
        const outputs = outputObjectTypes.map(factory(false)).filter(Boolean);
        return [...inputs, ...outputs];
    }
    static symbolName(name) {
        return (0, change_case_1.pascalCase)(`${name}Model`);
    }
    addImportDeclaration(moduleSpecifier, name, isDefault = false) {
        if (!this.#imports.has(moduleSpecifier)) {
            this.#imports.set(moduleSpecifier, {
                kind: ts_morph_1.StructureKind.ImportDeclaration,
                moduleSpecifier: moduleSpecifier,
                namedImports: []
            });
        }
        const moduleDeclaration = this.#imports.get(moduleSpecifier);
        if (isDefault) {
            moduleDeclaration.defaultImport = name;
        }
        else {
            const nameImports = moduleDeclaration.namedImports;
            if (!nameImports.includes(name)) {
                nameImports.push(name);
            }
        }
        return this;
    }
    toString() {
        return DmmfModel.symbolName(this.name);
    }
}
exports.DmmfModel = DmmfModel;
