// @tsed/prisma@7.8.1/lib/cjs/generator/domain/DmmfField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DmmfField = void 0;
const parseDocumentationAttributes_1 = require("../utils/parseDocumentationAttributes");
class DmmfField {
    model;
    field;
    schemaArg;
    constructor({ field, schemaArg, model }) {
        this.field = field;
        this.model = model;
        this.schemaArg = schemaArg;
    }
    get name() {
        return this.field.name;
    }
    get isRequired() {
        return this.field.isRequired;
    }
    get type() {
        return this.field.type;
    }
    get isList() {
        return this.field.isList;
    }
    get kind() {
        return this.field.kind;
    }
    get isNullable() {
        return this.schemaArg.isNullable;
    }
    get location() {
        return this.field.kind === "enum" ? "enumTypes" : this.field.kind === "object" ? "inputObjectTypes" : "scalar";
    }
    getAdditionalDecorators() {
        return (0, parseDocumentationAttributes_1.parseDocumentationAttributes)(this.field.documentation);
    }
    toString() {
        return this.name;
    }
}
exports.DmmfField = DmmfField;
