// @tsed/prisma@7.8.1/lib/cjs/generator/domain/ScalarTsTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScalarDecorators = exports.ScalarJsClasses = exports.ScalarTsTypes = exports.PrismaScalars = void 0;
var PrismaScalars;
(function (PrismaScalars) {
    PrismaScalars["String"] = "String";
    PrismaScalars["Boolean"] = "Boolean";
    PrismaScalars["Int"] = "Int";
    PrismaScalars["Float"] = "Float";
    PrismaScalars["BigInt"] = "BigInt";
    PrismaScalars["Decimal"] = "Decimal";
    PrismaScalars["DateTime"] = "DateTime";
    PrismaScalars["Json"] = "Json";
    PrismaScalars["Bytes"] = "Bytes";
})(PrismaScalars = exports.PrismaScalars || (exports.PrismaScalars = {}));
exports.ScalarTsTypes = {
    [PrismaScalars.String]: "string",
    [PrismaScalars.Boolean]: "boolean",
    [PrismaScalars.Int]: "number",
    [PrismaScalars.Float]: "number",
    [PrismaScalars.Decimal]: "number",
    [PrismaScalars.BigInt]: "bigint",
    [PrismaScalars.DateTime]: "Date",
    [PrismaScalars.Json]: "any",
    [PrismaScalars.Bytes]: "Buffer"
};
exports.ScalarJsClasses = {
    [PrismaScalars.String]: "String",
    [PrismaScalars.Boolean]: "Boolean",
    [PrismaScalars.Int]: "Number",
    [PrismaScalars.Float]: "Number",
    [PrismaScalars.Decimal]: "Number",
    [PrismaScalars.BigInt]: "BigInt",
    [PrismaScalars.DateTime]: "Date",
    [PrismaScalars.Json]: "Object",
    [PrismaScalars.Bytes]: "Buffer"
};
exports.ScalarDecorators = {
    [PrismaScalars.String]: [],
    [PrismaScalars.Boolean]: [],
    [PrismaScalars.Int]: [
        {
            name: "Integer",
            arguments: []
        }
    ],
    [PrismaScalars.Float]: [],
    [PrismaScalars.BigInt]: [],
    [PrismaScalars.Decimal]: [],
    [PrismaScalars.DateTime]: [{ name: "Format", arguments: ['"date-time"'] }],
    [PrismaScalars.Json]: [],
    [PrismaScalars.Bytes]: []
};
