// @tsed/platform-cache@7.8.1/lib/esm/services/PlatformCache.js
import { __decorate, __metadata } from "tslib";
import { isClass, isFunction, isString } from "@tsed/core";
import { Configuration, Inject, InjectorService, Module } from "@tsed/di";
import { deserialize, serialize } from "@tsed/json-mapper";
import { Logger } from "@tsed/logger";
const defaultKeyResolver = (args) => {
    return args.map((arg) => (isClass(arg) ? JSON.stringify(serialize(arg)) : arg)).join(":");
};
/**
 * @platform
 */
let PlatformCache = class PlatformCache {
    settings;
    injector;
    logger;
    #cache;
    get cache() {
        return this.#cache;
    }
    async $onInit() {
        const settings = this.settings.get("cache");
        if (settings) {
            this.#cache = await this.createCacheManager(settings);
            await this.injector.emit("$onCreateCacheManager", this.#cache);
        }
    }
    disabled() {
        return !this.settings.get("cache");
    }
    defaultKeyResolver() {
        return this.settings.get("cache.keyResolver", defaultKeyResolver);
    }
    defaultTtl() {
        return this.settings.get("cache.ttl");
    }
    calculateTTL(result, currentTtl) {
        const ttl = currentTtl === undefined ? this.defaultTtl() : currentTtl;
        return isFunction(ttl) ? ttl(result) : ttl;
    }
    async ttl(key) {
        if (this.cache && "store" in this.cache && this.cache.store.ttl) {
            return this.cache.store.ttl(key);
        }
    }
    wrap(key, fetch, ttl) {
        if (!this.cache) {
            return fetch();
        }
        return this.cache?.wrap(key, fetch, ttl);
    }
    async get(key, options = {}) {
        return deserialize(this.cache?.get(key), options);
    }
    async set(key, value, options) {
        await this.cache?.set(key, value, options?.ttl);
        return;
    }
    async getCachedObject(key) {
        try {
            return await this.get(key);
        }
        catch (er) {
            this.logger.error({
                event: "CACHE_ERROR",
                method: "getCachedObject",
                error: er
            });
        }
    }
    async setCachedObject(key, data, opts) {
        try {
            await this.set(key, {
                ...opts,
                data: JSON.stringify(data)
            }, {
                ttl: opts.ttl
            });
        }
        catch (er) {
            this.logger.error({
                event: "CACHE_ERROR",
                method: "setCachedObject",
                error: er
            });
        }
    }
    async del(key) {
        await this.cache?.del(key);
    }
    async reset() {
        // @ts-ignore
        await this.cache?.reset();
    }
    async keys(...args) {
        if (this.cache && "store" in this.cache && this.cache.store.keys) {
            return this.cache.store.keys(...args);
        }
        // istanbul ignore next
        return [];
    }
    async getMatchingKeys(patterns) {
        const [keys, { default: micromatch }] = await Promise.all([this.keys(), import("micromatch")]);
        return micromatch(keys, patterns);
    }
    async deleteMatchingKeys(patterns) {
        const keys = await this.getMatchingKeys(patterns);
        await Promise.all(keys.map((key) => this.del(key)));
        return keys;
    }
    async createCacheManager(settings) {
        const { caches, store = "memory", ttl, ...props } = settings;
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { multiCaching, caching } = await import("cache-manager");
        return caches?.length
            ? multiCaching(caches)
            : caching(this.mapStore(store), {
                ...props,
                ttl
            });
    }
    mapStore(store) {
        if (!isString(store) && "create" in store) {
            return store.create;
        }
        return store;
    }
};
__decorate([
    Configuration(),
    __metadata("design:type", Object)
], PlatformCache.prototype, "settings", void 0);
__decorate([
    Inject(),
    __metadata("design:type", InjectorService)
], PlatformCache.prototype, "injector", void 0);
__decorate([
    Inject(),
    __metadata("design:type", Logger)
], PlatformCache.prototype, "logger", void 0);
PlatformCache = __decorate([
    Module()
], PlatformCache);
export { PlatformCache };
