// @tsed/cli-core@4.2.4/lib/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Inquirer = exports.DITest = exports.Container = exports.InjectorService = exports.OverrideProvider = exports.UseOpts = exports.Opts = exports.Module = exports.LocalsContainer = exports.Configuration = exports.Value = exports.Constant = exports.Injectable = exports.registerProvider = exports.Inject = void 0;
const tslib_1 = require("tslib");
const inquirer_1 = tslib_1.__importDefault(require("inquirer"));
exports.Inquirer = inquirer_1.default;
tslib_1.__exportStar(require("@tsed/core"), exports);
tslib_1.__exportStar(require("@tsed/logger"), exports);
var di_1 = require("@tsed/di");
Object.defineProperty(exports, "Inject", { enumerable: true, get: function () { return di_1.Inject; } });
Object.defineProperty(exports, "registerProvider", { enumerable: true, get: function () { return di_1.registerProvider; } });
Object.defineProperty(exports, "Injectable", { enumerable: true, get: function () { return di_1.Injectable; } });
Object.defineProperty(exports, "Constant", { enumerable: true, get: function () { return di_1.Constant; } });
Object.defineProperty(exports, "Value", { enumerable: true, get: function () { return di_1.Value; } });
Object.defineProperty(exports, "Configuration", { enumerable: true, get: function () { return di_1.Configuration; } });
Object.defineProperty(exports, "LocalsContainer", { enumerable: true, get: function () { return di_1.LocalsContainer; } });
Object.defineProperty(exports, "Module", { enumerable: true, get: function () { return di_1.Module; } });
Object.defineProperty(exports, "Opts", { enumerable: true, get: function () { return di_1.Opts; } });
Object.defineProperty(exports, "UseOpts", { enumerable: true, get: function () { return di_1.UseOpts; } });
Object.defineProperty(exports, "OverrideProvider", { enumerable: true, get: function () { return di_1.OverrideProvider; } });
Object.defineProperty(exports, "InjectorService", { enumerable: true, get: function () { return di_1.InjectorService; } });
Object.defineProperty(exports, "Container", { enumerable: true, get: function () { return di_1.Container; } });
Object.defineProperty(exports, "DITest", { enumerable: true, get: function () { return di_1.DITest; } });
tslib_1.__exportStar(require("./interfaces"), exports);
tslib_1.__exportStar(require("./decorators"), exports);
tslib_1.__exportStar(require("./services"), exports);
tslib_1.__exportStar(require("./utils"), exports);
tslib_1.__exportStar(require("./CliCore"), exports);
require("./utils/patchCommander");
