// @tsed/prisma@7.8.1/lib/cjs/generator/utils/generateInterfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateInterfaces = void 0;
const tslib_1 = require("tslib");
const ts_morph_1 = require("ts-morph");
const path_1 = tslib_1.__importDefault(require("path"));
function generateInterfaces(project, baseDirPath) {
    const directory = project.createDirectory(path_1.default.resolve(baseDirPath, "interfaces"));
    const indexFile = directory.createSourceFile("index.ts", undefined, { overwrite: true });
    indexFile.addImportDeclaration({
        kind: ts_morph_1.StructureKind.ImportDeclaration,
        moduleSpecifier: "../client",
        namedImports: ["Prisma"]
    });
    indexFile
        .addModule({
        name: "",
        hasDeclareKeyword: true,
        declarationKind: ts_morph_1.ModuleDeclarationKind.Global
    })
        .addModule({
        name: "TsED"
    })
        .addInterface({
        name: "Configuration"
    })
        .addProperty({
        name: "prisma",
        hasQuestionToken: true,
        type: "Prisma.PrismaClientOptions"
    });
}
exports.generateInterfaces = generateInterfaces;
