// @tsed/schema@7.8.1/lib/cjs/constants/httpMethods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationMethods = exports.HTTP_METHODS = void 0;
exports.HTTP_METHODS = [
    "all",
    "checkout",
    "connect",
    "copy",
    "delete",
    "get",
    "head",
    "lock",
    "merge",
    "mkactivity",
    "mkcol",
    "move",
    "m-search",
    "notify",
    "options",
    "param",
    "patch",
    "post",
    "propfind",
    "propatch",
    "purge",
    "put",
    "report",
    "search",
    "subscribe",
    "trace",
    "unlock",
    "unsuscribe"
];
var OperationMethods;
(function (OperationMethods) {
    OperationMethods["ALL"] = "ALL";
    OperationMethods["GET"] = "GET";
    OperationMethods["POST"] = "POST";
    OperationMethods["PUT"] = "PUT";
    OperationMethods["PATCH"] = "PATCH";
    OperationMethods["HEAD"] = "HEAD";
    OperationMethods["DELETE"] = "DELETE";
    OperationMethods["OPTIONS"] = "OPTIONS";
    OperationMethods["CUSTOM"] = "CUSTOM";
})(OperationMethods = exports.OperationMethods || (exports.OperationMethods = {}));
