// @tsed/platform-params@7.8.1/lib/esm/pipes/ValidationPipe.js
import { __decorate, __metadata } from "tslib";
import { nameOf } from "@tsed/core";
import { Injectable, InjectorService } from "@tsed/di";
import { deserialize } from "@tsed/json-mapper";
import { getJsonSchema } from "@tsed/schema";
import { RequiredValidationError } from "../errors/RequiredValidationError.js";
import { ParamTypes } from "../domain/ParamTypes.js";
function cast(value, metadata) {
    try {
        return deserialize(value, {
            type: metadata.type
        });
    }
    catch (er) {
        return value;
    }
}
let ValidationPipe = class ValidationPipe {
    validator;
    constructor(injector) {
        const provider = injector.getProviders().find((provider) => nameOf(provider.token) === "AjvService");
        if (provider) {
            this.validator = injector.invoke(provider.token);
        }
    }
    coerceTypes(value, metadata) {
        if (value === undefined) {
            return value;
        }
        if (value === "null") {
            return null;
        }
        if (metadata.isArray) {
            return [].concat(value);
        }
        if (metadata.isPrimitive) {
            return cast(value, metadata);
        }
        return value;
    }
    skip(value, metadata) {
        return metadata.paramType === ParamTypes.PATH && !metadata.isPrimitive;
    }
    async transform(value, metadata) {
        if (!this.validator) {
            this.checkIsRequired(value, metadata);
            return value;
        }
        if (this.skip(value, metadata)) {
            return value;
        }
        value = this.coerceTypes(value, metadata);
        this.checkIsRequired(value, metadata);
        if (value === undefined) {
            return value;
        }
        const schema = getJsonSchema(metadata, {
            customKeys: true
        });
        await this.validator.validate(value, {
            schema,
            type: metadata.isClass ? metadata.type : undefined,
            collectionType: metadata.collectionType
        });
        return value;
    }
    checkIsRequired(value, metadata) {
        if (metadata.isRequired(value)) {
            throw RequiredValidationError.from(metadata);
        }
        return true;
    }
};
ValidationPipe = __decorate([
    Injectable({
        type: "validator"
    }),
    __metadata("design:paramtypes", [InjectorService])
], ValidationPipe);
export { ValidationPipe };
