// @tsed/cli-core@4.2.4/lib/esm/utils/getPackageJson.js
import readPkgUp from "read-pkg-up";
import { dirname, join } from "path";
import Fs from "fs-extra";
function useReadPkgUp(configuration) {
    return !(process.argv.includes("init") && !Fs.existsSync(join(String(configuration.project?.rootDir), "package.json")));
}
function getEmptyPackageJson(configuration) {
    return {
        name: configuration.name,
        version: "1.0.0",
        description: "",
        scripts: {},
        dependencies: {},
        devDependencies: {}
    };
}
export function getPackageJson(configuration) {
    if (useReadPkgUp(configuration)) {
        const result = readPkgUp.sync({
            cwd: configuration.project?.rootDir
        });
        if (result && result.path) {
            const pkgPath = dirname(result.path);
            configuration.set("project.root", pkgPath);
            const pkg = Fs.readJsonSync(result.path, { encoding: "utf8" });
            return { ...getEmptyPackageJson(configuration), ...pkg };
        }
    }
    return getEmptyPackageJson(configuration);
}
